src/cargo/core/compiler/custom_build.rs

```diff
@@ -643,26 +647,97 @@ impl BuildOutput {
+                    if !new_syntax {
+                        bail!("invalid output in {}: `{}`\n\
+                            Expected a line with `cargo::metadata=KEY=VALUE`, but found `cargo:metadata=KEY=VALUE`.\n\
+                            See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                            for more information about build script outputs.", whence, line,
+                        )
+                    }
+                    // Reserved keys should not be used with the `cargo::metadata` syntax.
+                    if new_syntax && RESERVED_KEYS.contains(&a) {
+                        bail!("invalid output in {}: `{}`\n\
+                            The reserved key `{}` cannot be used in a `cargo::metadata` line. Please use `cargo::{}=VAlUE` instead.\n\
+                            See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                            for more information about build script outputs.", whence, line, a, a
+                        )
+                    }
+                    (a, b.trim_end().to_owned())
+                },
+                (Some(a), Some(b), None) => {
+                    // Only reserved keys should be used with the `cargo::` syntax.
+                    // Otherwise, it should be `cargo::metadata=KEY=VALUE`.
+                    if new_syntax && !RESERVED_KEYS.contains(&a) {
+                        bail!("invalid output in {}: `{}`\n\
+                            Expected a line with `cargo::metadata=KEY=VALUE` but it did not have the `metadata=` part.\n\
+                            See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                            for more information about build script outputs.", whence, line
+                        )
+                    }
+                    (a, b.trim_end().to_owned())
+                },
+                (Some(a),Some(b),Some(c)) => {
+                    // Only reserved keys should be used with the `cargo::` syntax.
+                    // Otherwise, it should be `cargo::metadata=KEY=VALUE`.
+                    if new_syntax && !RESERVED_KEYS.contains(&a) {
+                        bail!("invalid output in {}: `{}`\n\
+                            Expected a line with `cargo::metadata=KEY=VALUE` but it did not have the `metadata=` part.\n\
+                            See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                            for more information about build script outputs.", whence, line
+                        )
+                    }
+                    // For instance, `cargo:rustc-link-search=[KIND=]PATH` or `cargo::rustc-link-search=[KIND=]PATH`.
+                    (a, format!("{}={}", b, c).trim_end().to_owned())
+                }
```

Comment:
This feels like its a lot of complexity and it looks like there are error cases that shouldn't exist like `cargo::metadata`

Couldn't we just update `match key` to have a `"metadata" if new_syntax` and `_ if new_syntax => ... error ..`?

We'd of course need to make sure the `metadata` case is first (to make sure its discoverable when people add new keys) and document that all new keys need to be guarded with that expression.

Reply:
> Couldn't we just update `match key` to have a `"metadata" if new_syntax` and `_ if new_syntax => ... error ..`?

If we do it in this way maybe it is more complicated. Because the new syntax has three parts or two parts and the old syntax has two parts. So it is difficult to only use a simple match to cover it.


> We'd of course need to make sure the `metadata` case is first (to make sure its discoverable when people add new keys) and document that all new keys need to be guarded with that expression.

Do you mean the new reserved key also needs to be like `cargo::metadata=new-key=VALUE`?

Reply:
> If we do it in this way maybe it is more complicated. Because the new syntax has three parts or two parts and the old syntax has two parts. So it is difficult to only use a simple match to cover it.

The point is to make the code not care about two vs three parts

One advantage of you approach is it makes it more likely to Do The Right thing when new keys are added.

Maybe a mixed approach, something like:
```rust
let (key, value) = if let Some(data) = line.strip_prefix("cargo:") {
        parse_directive(directive)
} else if let Some(directive) = line.strip_prefix("cargo:") {
    // Due to backwards compatibility, no new keys can be added to this old format
    const SUPPORTED_PREFIXES: &[&str] = [
        "rustc-flags=",
        // ...
    ];
    if SUPPORTED_PREFIXES.iter().any(|prefix| data.starts_with(prefix) {
        parse_directive(directive)
    } else {
        ("metadata", directive)
    }
} else {
    continue;
};

// This will rewrite paths if the target directory has been moved.
let value = value.replace(

// ...

fn parse_directive(directive: &str) -> CargoResult<(&str, &str)> {
    data.split_once('=').ok_or_else(|| {
        // ... error
    })
        // ... error
    };
}
```
And `match key {` would need to be updated
- The current `_` case would need to instead be `"metadata"`
- A new `_` case is needed that would error, saying that they key is unsupported

Advantages
- The old code path is isolated from the rest of the behavior
- New keys can be added without accidentally breaking compatibility
- We only error in the cases we need to
- It looks like it'll be smaller and cleare

Reply:
I updated it. I followed your approach. But I don't match the `metadata` in the `match key {` part, because if we do that we need to parse the meta key-value pair in `metadata  => {}`.
I always parse key values before we want to match them. Then we do not need to special handle the metadata key. We can directly put it into the metadata array. And in the future, if we want to add a new key, then directly match it in the `match key {` part. You can check out: https://github.com/rust-lang/cargo/pull/12201/commits/9e464ed0a44e90bfd8e7f49dbd23df46c16f6a40

Thanks for your suggestion! It is much easier to add a new key in the future and the code is clearer.

Reply:
> I updated it. I followed your approach. But I don't match the metadata in the match key { part, because if we do that we need to parse the meta key-value pair in metadata  => {}.

Could you expand on this?  Its not quite making sense.

Using the `metadata` key is important because
- It clarifies the code.  Having the "key" be both the directive and the metadata-key is confusing
- The current design still prevents adding new keys with the new syntax
- It isolates knowledge of the old syntax, especially when correctly allowing new keys to be added

Reply:
> * The current design still prevents adding new keys with the new syntax

It makes adding new keys easier, if you want to add a new key, just add it to the `match key` part. You don't need to change any other code.

Another reason I didn't use the `metadata` key is that we need to consider the old syntax.

For example, in the old syntax, if we use `cargo::some-metadata-key=some-metadata-value`, it would be added to the metadata map. 

If we use the `metadata` key, how to deal with this kind of situation?  Do we need to add more code to unify this behavior? 

Maybe we need to do it like this:
```rust
let (key, value) = if let Some(data) = line.strip_prefix("cargo:") {
    if SUPPORTED_PREFIXES.iter().any(|prefix| data.starts_with(prefix) {
        parse_directive(directive)
    } else {
        ("metadata", directive)
    }
} else if let Some(directive) = line.strip_prefix("cargo:") {
    // Due to backwards compatibility, no new keys can be added to this old format
    const SUPPORTED_PREFIXES: &[&str] = [
        "rustc-flags=",
        // ...
    ];
    if SUPPORTED_PREFIXES.iter().any(|prefix| data.starts_with(prefix) {
        parse_directive(directive)
    } else {
        ("metadata", directive)
    }
} else {
    continue;
};
```

Reply:
If we really want to do it. Here are some problems:

1. We can not simply return `("metadata", directive)`, otherwise we need to handle `cargo::foo=bar` in the `match "metadata"` part. Maybe it does not make sense.
2. For `match "metadata"`, we still need to couple the old syntax and new syntax together. For example, we need to handle `cargo:foo=bar` and `cargo::metadata=foo=bar` together.
3. We don't really have a `_ => {}` statement for `match key` because we cover all situations in `match "metadata"` and other branches.

Reply:
I tried it in https://github.com/hi-rustin/cargo/commit/cce5bcbca22cb77fc25e2cab93d9a1f450b20096

Could you please take a look? Thanks! But if we want to add a new key, it is a little bit difficult to handle it in the `match key`.